operator: Init MetalK8s Operator

TeddyAndrieux opened the pull request:
For reviewer: You should review commits one by one as some of the commits is just "some code" generated by `operator-sdk`

---

- Init operator and ClusterConfig CRD :heavy_check_mark: 
- Operator image build :heavy_check_mark:
- Operator manifests builds :heavy_check_mark: 
- Operator go linting in CI :heavy_check_mark: 
- Operator go codegen in doit :heavy_check_mark: 
- Operator go test in CI :heavy_check_mark: 
- Operator deployment :heavy_check_mark:
- Create a default ClusterConfig object at operator startup (and delete any other ClusterConfig objects) :heavy_check_mark:
- Add a test to ensure the default ClusterConfig get recreated and "extra" ClusterConfig get deleted :heavy_check_mark: 
- Add changelog entry :heavy_check_mark: 

NOTE: We do not add any Webhook to ensure that you cannot create/delete ClusterConfig objects in this PR as it needs certificates and we may want to use Cert-manager to manage those certs, so it will be done in another PR

TeddyAndrieux committed:
build: Remove no longer needed go constant

TeddyAndrieux committed:
build,operator,tests: Add operator deployment

Add build for MetalK8s operator image and for operator manifests using
kustomize with make

```
make metalk8s
```

Also add a sanity test to ensure the MetalK8s operator is properly
deployed

TeddyAndrieux committed:
build: Add linting and codegen for MetalK8s operator in doit

TeddyAndrieux committed:
eve: Add go unit tests for MetalK8s operator

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @chengyanjin

* @thomasdanan

* @alexandre-allard-scality

* @gdemonet

* @teddyandrieux

* @jbwatenbergscality

* @lucieleonard

* @nicolast

TeddyAndrieux committed:
operator: Create the default ClusterConfig from the Operator

Also add the logic to only reconcile this object and delete others

TeddyAndrieux committed:
tests: Add simple test about ClusterConfig creation/deletion

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @lucieleonard

* @teddyandrieux

* @chengyanjin

* @gdemonet

* @jbwatenbergscality

* @thomasdanan

* @alexandre-allard-scality

TeddyAndrieux committed:
changelog: Add changelog entry for `metalk8s-operator`

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/init-cluster-config-operator` into
`development/123.0`
* [w/124.0/improvement/init-cluster-config-operator](https://github.com/scality/metalk8s/compare/development/124.0...w/124.0/improvement/init-cluster-config-operator) will be merged into `development/124.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @lucieleonard

* @teddyandrieux

* @chengyanjin

* @gdemonet

* @jbwatenbergscality

* @thomasdanan

* @alexandre-allard-scality

JBWatenbergScality commented on operator/PROJECT:
```diff
@@ -0,0 +1,9 @@
+  manifests.sdk.operatorframework.io/v2: {}
+  scorecard.sdk.operatorframework.io/v2: {}
+projectName: operator
+repo: github.com/metalk8s/operator
```
nit: Shall not we simply use `github.com/scality/metalk8s` instead of `github.com/metalk8s/operator` as we don't really own the repo `github.com/metalk8s/operator` and I think ideally this info should match the repo holding the source code of the operator as a best practice (some tooling for doc may rely on this info to generate links or so...)

TeddyAndrieux replied:
Hmm, I'm not sure, cause we have multiple operators in the same repo and the "root" for the operator is this directory. And to me this repo is the "URL" to the operator go I think, no ?

I don't know about the best practice and not have any strong opinion here, just today we used the same approach for the storage-operator (`repo: github.com/scality/metalk8s/storage-operator`)
Any opinion on this @gdemonet ?

gdemonet replied:
Yes I think we should use the "real" URL, so none of `github.com/scality/metalk8s` or `github.com/metalk8s/operator` are valid :innocent: 

```suggestion
repo: github.com/scality/metalk8s/operator
```

TeddyAndrieux replied:
Oh ! Good catch indeed I forgot the `scality`, but I don't see why `github.com/scality/metalk8s` is not valid

TeddyAndrieux replied:
Fixed

JBWatenbergScality commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+
+// Trigger a reschedule after a short delay.
+func delayedRequeue(err error) (ctrl.Result, error) {
+	delay := 10 * time.Second
```
nit: shall we instead of a fixed retry time use an exponential backoff approach to avoid overload the API server with failures when something wrong is already happening ?

TeddyAndrieux replied:
Yes could make sense, for "usual" operator, I agree, but here we talk about an operator that can handle only one single object so ... I'm not sure it bring a lot of value here

TeddyAndrieux replied:
So I will let it like that for the moment and we will see later if we may do something a bit smarter to compute the delay

TeddyAndrieux replied:
I did some research and the controller-runtime is automatically implementing a retry back-off for us

JBWatenbergScality commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+}
+
+// Trigger a reschedule after a short delay.
+func delayedRequeue(err error) (ctrl.Result, error) {
+	delay := 10 * time.Second
+	return ctrl.Result{Requeue: err == nil, RequeueAfter: delay}, err
+}
+
+// Trigger a reschedule as soon as possible.
+func requeue(err error) (ctrl.Result, error) {
+	return ctrl.Result{Requeue: err == nil}, err
+}
```
nit: maybe those methods should try to update the status of the CR to notify about a failure ?

TeddyAndrieux replied:
Yes, you're right maybe if we have an error we want to put it in the object status :thinking:

gdemonet replied:
This shouldn't be part of the "requeue" methods. Deciding whether to show an error in the managed object is part of the reconciliation logic of a specific controller, requeuing is just sending information to the controller manager.

TeddyAndrieux replied:
No strong opinion here, it could be an option but indeed we may have some cases where we just want to requeue and not update the object

TeddyAndrieux replied:
I will let it like that for the moment and we will see when we will add more logic to the operator if we want to update the status or not

JBWatenbergScality commented on tests/post/features/operator.feature:
```diff
@@ -0,0 +1,13 @@
+        When we create an extra 'test-cc' ClusterConfig
+        Then the 'test-cc' ClusterConfig get deleted
+
+    Scenario: Deletion of the default ClusterConfig
+        Given the Kubernetes API is available
+        And pods with label 'app.kubernetes.io/name=metalk8s-operator' are 'Ready'
+        When we delete the 'default' ClusterConfig
+        Then the 'default' ClusterConfig get automatically recreated
```
nit: With this approach we are constraining the use of Metalk8s operator only with Metalk8s (which is likely fine for now) however I think we might want to use it to install adds-on easily on other K8s distributions, to be discussed. 
The test is plenty valid for Metalk8s context through, but I let the comment here as a base to discuss.

TeddyAndrieux replied:
I'm not sure to get your point here.

I agree that such operator could be used outside of "metalk8s core" at some point, but I don't get why this approach could be an issue for this ?

NOTE: This test will be updated at some point when we will have a proper Webhook to "forbib" object creation and deletion

gdemonet replied:
Are you suggesting we shouldn't use a `ClusterConfig` object when running in other K8s distributions @JBWatenbergScality? It's debatable, and adding support for this could be done later IMO.

JBWatenbergScality replied:
I think we can keep it like this for now. 
For the long term goal I was thinking about using the same ClusterConfig but without the "core" part, hence in such cases we should be able to delete the resource. But anyway let's start with this implementation and iterate over it later on

TeddyAndrieux committed:
operator: Init MetalK8s operator

Init the MetalK8s operator using operator-sdk
```
operator-sdk init --domain metalk8s.scality.com --repo github.com/scality/metalk8s/operator
```

TeddyAndrieux committed:
operator: Create the ClusterConfig API and resource

Using operator-sdk
```
operator-sdk create api --version v1alpha1 --kind ClusterConfig --resource --controller --namespaced=False
make generate
make manifests
```

TeddyAndrieux committed:
operator: Update a bit the ClusterConfig spec

Update the ClusterConfig spec to add "WorkloadPlane" field and
regenerate the CRD

```
make generate
make manifests
```

gdemonet commented on operator/api/v1alpha1/clusterconfig_types.go:
```diff
@@ -23,13 +23,17 @@ import (
 
-	// Foo is an example field of ClusterConfig. Edit clusterconfig_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// Information about the Workload Plane.
+	// +optional
+	WorkloadPlane WorkloadPlaneSpec `json:"workloadPlane,omitempty"`
```
:thinking:

We have received requests for handling more than a single "workload plane network", I'm wondering if the assumption we took about the network layout (1 control plane + 1 workload plane) wasn't a bit too restrictive... I guess we'd still need to special-case the control plane, but I think we'll end up changing this assumption to something more generic (i.e. managing an arbitrary number of networks, for which we'll have ingress controllers, and out of which a single one can bear the control plane services).

It's likely too soon to jump on this, but I bet we'll have to rework the CR structure to handle this differently.

TeddyAndrieux replied:
I'm not sure, to me we will keep 1 network for Control Plane and 1 network for Workload Plane, but we may expose Ingresses on other networks/IPs.

Why we would need to change the CR layout, to me with the same layout as Bootstrap config we are fine

gdemonet replied:
My point is that the assumptions about what services should use which network could be wrong (e.g. don't use the public-facing workload plane for the overlay network, or don't expose the control-plane-ingress-controller on the control plane network), and this will likely become easier to handle if we don't encode such assumptions in the CR structure.

The fact that BootstrapConfiguration also has these assumptions encoded would also become a problem, IMO.

Anyway, let's not jump to conclusions on this topic, we'll try to think about it more in the future. And changes to the CR schema are possible, so not much to worry about.

gdemonet commented on buildchain/buildchain/image.py:
```diff
@@ -280,6 +280,19 @@ def _local_image(name: str, **kwargs: Any) -> targets.LocalImage:
+            .isoformat(),
+            "VCS_REF": constants.GIT_REF or "<unknown>",
+            "METALK8S_VERSION": versions.VERSION,
+            "VERSION": "latest",
```
Why are we doing this? Couldn't we just tag our operators with the project version? (I see we actually use the version in the image tag, but this label should reflect the same value I think)

TeddyAndrieux replied:
TBH I don't really know, I did the same as for the storage operator, I think you should ask it to @gdemonet who did it for storage operator https://github.com/scality/metalk8s/commit/fdeae9674719c683e4955de7f081da6087fc32bc#diff-cdf9db2f127c381762a567b95b0e86870b5ef70f25d5e38750d5294bc3597044 :stuck_out_tongue_winking_eye:

gdemonet replied:
:trollface:

gdemonet replied:
OK, let's not change storage-operator right now, but can you change this line to:

```suggestion
            "VERSION": versions.VERSION,
```

please?

TeddyAndrieux replied:
Sure I will do it

TeddyAndrieux replied:
Done

gdemonet commented on operator/deploy/manifests.yaml:
```diff
@@ -0,0 +1,227 @@
+---
+apiVersion: v1
+data:
+  controller_manager_config.yaml: "apiVersion: controller-runtime.sigs.k8s.io/v1alpha1\nkind: ControllerManagerConfig\nhealth:\n  healthProbeBindAddress: :8081\nmetrics:\n  bindAddress: 127.0.0.1:8080\nwebhook:\n  port: 9443\nleaderElection:\n  leaderElect: true\n  resourceName: 02367e3e.metalk8s.scality.com\n#   leaderElectionReleaseOnCancel defines if the leader should step down volume \n#   wh ...
```
Not super readable :disappointed:

TeddyAndrieux replied:
It's auto-generated, but yes indeed

gdemonet commented on operator/config/metalk8s/deploy_patch.yaml:
```diff
@@ -0,0 +1,23 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: not-important
```
Do we still need it if it's not relevant?

TeddyAndrieux replied:
Unfortunately yes, `name` is mandatory but not used

gdemonet commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -33,6 +37,13 @@ type ClusterConfigReconciler struct {
+// the object is created by the operator and it's the only one that will be managed
+var (
+	log          = logf.Log.WithName("clusterconfig-controller")
+	instanceName = "default"
```
(nit) Not sure about the name, I'd rather use "main" :thinking: WDYT? Also, isn't it a `const` rather than a `var`?

TeddyAndrieux replied:
Right, I will change it

TeddyAndrieux replied:
Done

gdemonet commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+// Trigger a reschedule after a short delay.
+func delayedRequeue(err error) (ctrl.Result, error) {
+	delay := 10 * time.Second
+	return ctrl.Result{Requeue: err == nil, RequeueAfter: delay}, err
```
Just for my own understanding, if `err != nil`, does it mean we will still `RequeueAfter`?

TeddyAndrieux replied:
I did not tested this, this function comes from the storage-operator, from the documentation if `Requeue` is False he will not be requeued so I think no but I will do a test

TeddyAndrieux replied:
So I did some tests and research on the internet, and it appears that if there is an error in "Reconcile" return, it will be automatically requeued no matter the "Result" content AND the Controller automatically handle the retry backoff.

So I will update a bit the functions

TeddyAndrieux replied:
Done

gdemonet commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+			)
+			return delayedRequeue(err)
+		}
+		reqLogger.Info("deleting extra ClusterConfig, consider updating the default one")
```
Would be better to print the actual `instanceName` in this message I believe

TeddyAndrieux replied:
Hmm yes indeed, I will fix it

TeddyAndrieux replied:
Done

gdemonet commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+		Spec:       metalk8sscalitycomv1alpha1.ClusterConfigSpec{},
+	}
+
+	err := client.Create(context.Background(), instance)
```
Is it safe to use such a context when attempting to interact with K8s API? I guess we'd end up getting a timeout from k-a or the network layer if something is really wrong?

TeddyAndrieux replied:
This function is only called at operator startup, but yes we might be able to use another context, I will see

TeddyAndrieux replied:
Done

gdemonet commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+	client := mgr.GetClient()
+
+	instance := &metalk8sscalitycomv1alpha1.ClusterConfig{
+		ObjectMeta: metav1.ObjectMeta{Name: instanceName},
```
Could be worth adding a finalizer, just to prevent mistakes (which could be quite troublesome if there were a lot of customizations in the object)

TeddyAndrieux replied:
I'm not sure.

We could add a finalizer that we never remove but the fact is AFAIK we cannot "cancel" a deletion, so even if we have a finalizer, if the user deletes the object, it will be marked as deleted and will stay in this state

So ... I don't know, and to me, I see this as, a user deleting the ClusterConfig object so we basically "reset" to the default config (today the default config is empty but maybe at some point we will have some "good" default), make sense no ?

And at some point, we will have webhook to forbid it

gdemonet replied:
I think we can cancel a deletion by removing `metadata.deletionTimestamp` and `metadata.deletionGracePeriodSeconds`, but yeah OK, since we will add the webhook I guess that's fine. And since we don't want to handle deletion, adding a finalizer doesn't make much sense.

For "resetting to defaults", I think updating the object to have `spec: null` could be enough, no need to delete.

TeddyAndrieux replied:
AFAIK, we cannot remove those fields.

gdemonet commented on operator/controllers/clusterconfig_controller.go:
```diff
@@ -47,15 +58,72 @@ type ClusterConfigReconciler struct {
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if req.Name == instanceName {
+				panic(fmt.Errorf(
+					"%s ClusterConfig object should not be deleted", req.Name,
+				))
```
Worth adding a comment about the operator behavior here (either directly in the message, or at least in the code): it's not obvious that the panic will lead to a restart (because, K8s) followed by a re-creation of the main ClusterConfig on startup.

TeddyAndrieux replied:
Ok I will add a comment

TeddyAndrieux replied:
Done

gdemonet commented on operator/.gitignore:
```diff
@@ -0,0 +1,25 @@
+
+# Kubernetes Generated files - skip generated files, except for vendored files
+
+!vendor/**/zz_generated.*
```
? I don't understand, this line should imply we ignore generated files, but they are still part of some following commits :thinking:

TeddyAndrieux replied:
This gitignore come from operator-sdk.

To me it's the opposite `!`, we do not ignore the `zz_generated.*` files.

I think it's either a leftover in operator-sdk because they usually git ignore the `vendor` directory, either just here to ensure we do not git ignore those specific files

gdemonet replied:
Right, that's strange, the comment is quite misleading.

gdemonet reviewed (dismissed):
LGTM :ok_hand:

TeddyAndrieux commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b652d377842d07c8159646bd1c3572ee8b688467) did not succeed in branch improvement/init-cluster-config-operator.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/124.0`



The following branches will **NOT** be impacted:

* `development/123.0`
* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/124.0`



The following branches have **NOT** changed:

* `development/123.0`
* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
